Cython/Compiler/ExprNodes.py

```diff
@@ -790,7 +815,9 @@ def generate_disposal_code(self, code):
                 # postponed from self.generate_evaluation_code()
                 self.generate_subexpr_disposal_code(code)
                 self.free_subexpr_temps(code)
-            if self.result():
+            if (not self.type.is_cpp_class  # check this before calling result()
```

Comment:
I'd rather `and`-merge the `self.result()` condition into the two conditions below. No need to call it even unless we find a type for which we need to do something.